ui/pages/device_code.tsx

```diff
@@ -1,25 +1,83 @@
 import type { NextPage } from "next";
-import React from "react";
+import axios, { AxiosError } from "axios";
+import { useRouter } from "next/router";
+import React, { FormEvent, useCallback, useEffect, useState } from "react";
 import PageLayout from "../components/PageLayout";
-import { Button, Input } from "@canonical/react-components";
+import { Button, Form, Input } from "@canonical/react-components";
+
+export interface Response {
+  data: {
+    redirect_to?: string;
+  };
+}
+
+async function acceptUserCode(userCode: string, challenge: string) {
+  return axios
+    .put("/api/device?device_challenge=" + String(challenge), {
+      user_code: userCode,
+    })
+    .then(({ data }: Response) => {
+      if (data.redirect_to) {
+        window.location.href = data.redirect_to;
+      }
+    });
+}
 
 const DeviceCode: NextPage = () => {
+  const router = useRouter();
+  const { device_challenge: challenge, user_code: code } = router.query;
+  const [errorMessage, setErrorMessage] = useState("");
+
+  useEffect(() => {
+    // If the router is not ready yet, or we already have a flow, do nothing.
+    if (!router.isReady) {
+      return;
+    }
+  }, [code, errorMessage]);
+
+  const handleSubmit = useCallback(
+    (event: FormEvent<HTMLFormElement>) => {
+      event.preventDefault();
+
+      const formData = new FormData(event.currentTarget);
+      const userCode = formData.get("code");
+      acceptUserCode(String(userCode), String(challenge)).catch(
+        (error: AxiosError) => {
+          // TODO(nsklikas): Refactor when proper error handling is
+          // implemented in the backend
+          if (error.response?.status == 400) {
+            setErrorMessage(
+              "The code is either invalid, expired or already used",
```

Comment:
Isn't the `error.response.data.error_description` available to be used here?

Reply:
The backend will return an error saying `The user_code provided is either invalid, expired or already used` (it's not a json after all, to be consistent with the rest of the endpoints).

We do a very bad job at how we expose errors in the backed, we will need to refactor it in the future and we can revisit this then